File: sort_msm.py

import numpy as np


class PlayerTeam():
    def __init__(self, id:int, name:str, location:tuple[float, float]):
        self.id = id
        self.name = name
        self.location = location
        self.team = []
        self.friends = []
        self.distances = []
        self.rankings = []
        self.min_roster_size = None
        self.max_roster_size = None

    def sort_distances(self):
        self.distances = sorted(self.distances, key=lambda x: x[0])
    
    def add_to_team(self, team_id) -> bool:
        self.team.append(team_id)

    def room_on_team(self) -> bool:
        if not self.max_roster_size:
            return False
        if len(self.team) < self.max_roster_size:
            return True
        return False


def med_school_match(players: list[PlayerTeam], teams:list[PlayerTeam]):
    '''
    Match players to teams.  Starting with a player's top ranked team from the players.rankings list, 
    see if that team has that player ranked first, and if the team has room.  Loop over all players
    to match top ranks.  
    Continue to iterate, going deeper into the team's ranking and adding that player if there is room.
    If a player's first ranked team is full, remove it from the player's rankings, and so on until
    all players are assigned to a team.
    '''
    for player in players:
        player.rankings = [x[1] for x in player.distances]
    # Create a lookup for teams by their IDs
    team_dict = {team.id: team for team in teams}
    team_match_depth = 1
    # Create a dictionary to track which players are assigned
    assigned_players = set()
    # Continue matching until all players are assigned
    while len(assigned_players) < len(players):
        for player in players:
            if player.team:
                continue  # Skip already assigned to a team
            # Try to assign the player based on their preferences
            preferred_team_id = player.rankings[0]
            team = team_dict.get(preferred_team_id)
            # Check if the team exists and has space
            if len(team.team) == team.max_roster_size:  # team is full, move on to the next team
                player.rankings.remove(preferred_team_id)
                continue
            team_player_rankings = [x[1] for x in team.distances][:team_match_depth]
            if player.id in team_player_rankings:
                # Assign the player to the team
                player.add_to_team(team.id)
                team.add_to_team(player.id)
                assigned_players.add(player.id)
                continue
        team_match_depth += 1
    return 

File: sort_msm.js

function medSchoolMatch(players, teams) {
    // Initialize player rankings from distances
    players.forEach(player => {
        player.rankings = player.distances.map(distance => distance[1]);
    });

    // Create a lookup for teams by their IDs
    const teamDict = Object.fromEntries(teams.map(team => [team.id, team]));
    let teamMatchDepth = 1;
    const assignedPlayers = new Set();

    // Continue matching until all players are assigned
    while (assignedPlayers.size < players.length) {
        players.forEach(player => {
            if (player.team.length > 0) {
                return; // Skip players already assigned to a team
            }

            // Try to assign the player based on their preferences
            const preferredTeamId = player.rankings[0];
            const team = teamDict[preferredTeamId];

            if (!team) return; // Skip if team not found

            // Check if the team has space
            if (team.team.length === team.maxRosterSize) {
                // Team is full; remove it from player's rankings
                player.rankings.shift();
                return;
            }

            // Check if the player is within the team's top preferences
            const teamPlayerRankings = team.distances
                .slice(0, teamMatchDepth)
                .map(distance => distance[1]);

            if (teamPlayerRankings.includes(player.id)) {
                // Assign the player to the team
                player.addToTeam(team.id);
                team.addToTeam(player.id);
                assignedPlayers.add(player.id);
            }
        });

        teamMatchDepth += 1;
    }

    return;
}


// const fs = require('fs'); // Import the filesystem module for file operations

// function savePlayerTeamsToJSON(playerTeams, filename) {
//     // Convert PlayerTeam objects to plain objects for JSON serialization
//     const data = playerTeams.map(playerTeam => ({
//         id: playerTeam.id,
//         name: playerTeam.name,
//         location: playerTeam.location,
//         team: playerTeam.team,
//         friends: playerTeam.friends,
//         distances: playerTeam.distances,
//         rankings: playerTeam.rankings,
//         minRosterSize: playerTeam.minRosterSize,
//         maxRosterSize: playerTeam.maxRosterSize
//     }));

//     // Serialize to JSON and write to the specified file
//     fs.writeFileSync(filename, JSON.stringify(data, null, 2), 'utf-8');
// }

// function loadPlayerTeamsFromJSON(filename) {
//     // Read the file content
//     const data = fs.readFileSync(filename, 'utf-8');
    
//     // Parse the JSON data and convert to PlayerTeam instances
//     const parsedData = JSON.parse(data);
//     return parsedData.map(item => {
//         const playerTeam = new PlayerTeam(item.id, item.name, item.location);
//         playerTeam.team = item.team;
//         playerTeam.friends = item.friends;
//         playerTeam.distances = item.distances;
//         playerTeam.rankings = item.rankings;
//         playerTeam.minRosterSize = item.minRosterSize;
//         playerTeam.maxRosterSize = item.maxRosterSize;
//         return playerTeam;
//     });
// }


// // Example usage:
// const teams = [
//     new PlayerTeam(1, "Team A", [34.0522, -118.2437]),
//     new PlayerTeam(2, "Team B", [40.7128, -74.0060])
// ];

// // Customize properties for demonstration
// teams[0].team.push(101, 102);
// teams[1].minRosterSize = 5;
// teams[1].maxRosterSize = 10;

// savePlayerTeamsToJSON(teams, 'player_teams.json');
